src/libraries/System.Net.Security/src/System/Net/Security/Pal.OSX/SafeDeleteSslContext.cs

```diff
@@ -158,13 +174,24 @@ private unsafe int WriteToConnection(void* connection, byte* data, void** dataLe
             ulong length = (ulong)*dataLength;
             Debug.Assert(length <= int.MaxValue);
 
+            if (Interlocked.Exchange(ref _pendingOutput, NativeIOPending) == Disposed)
```

Comment:
Should this use CompareExchange?

Reply:
We really want to prevent Dispose from running and we want to detect in IO functions what dispose was called to run deferred cleanup. We don't really need to distinguish what is running.

Now looking at this again. It is not safe to manipulate the ArrayBuffer concurrently. We would need lock here to prevent corruption. But the code works because the general flow is like:
- we append received frame to queue.
- we call native code and that consumers the bytes
- it can write more data to the other queue
- we check the queue when finish and we send data to underlying stream

If we go with ReservableArrayBuffer this would turn into counter as you said. 
I like it better than the custom state machine so I'll make change as you recommended.